File: source/native/src/expression/Expression.cpp
#include <cmath>
#include "Expression.hpp"

// Expression

Expression::Expression(string type_): type(type_)
{
}

Value Expression::evaluate(Context* context)
{
    return Value(0);
}

string Expression::toString()
{
    return "";
}

BottomExpression::BottomExpression(): Expression("(bottom)")
{
}

Value BottomExpression::evaluate(Context* context)
{
    return Value();
}

string BottomExpression::toString()
{
    return "";
}

// LiteralExpression

LiteralExpression::LiteralExpression(Value value_): Expression("(literal)"), value(value_)
{
}

Value LiteralExpression::evaluate(Context* context)
{
    return value;
}

string LiteralExpression::toString()
{
    if (value.isString()) {
        return "\"" + value.stringValue() + "\"";
    }

    return value.stringValue();
}


// VariableExpression

VariableExpression::VariableExpression(string name_): Expression("(variable)"), name(name_)
{
}

Value VariableExpression::evaluate(Context* context)
{
    return context->get(name);
}

string VariableExpression::toString()
{
    return name;
}

// DynamicVariableExpression

DynamicVariableExpression::DynamicVariableExpression(string name_, shared_ptr<Expression> argument_): Expression("(dynamic_variable)"), baseName(name_), argument(argument_)
{
}

string DynamicVariableExpression::getName(Context* context)
{
    return baseName + "." + std::to_string(argument->evaluate(context).intValue());
}

Value DynamicVariableExpression::evaluate(Context* context)
{
    return context->get(getName(context));
}

string DynamicVariableExpression::toString()
{
    return baseName + "(" + argument->toString() + ")";
}

// FunctionCalExpression

FunctionCallExpression::FunctionCallExpression(string name_, vector<shared_ptr<Expression>> arguments_): Expression("(call)"), name(name_), arguments(arguments_)
{
}

Value FunctionCallExpression::evaluate(Context* context)
{
    vector<shared_ptr<Expression>> evaluatedArguments;
    for (int i = 0; i < arguments.size(); i++) {
        evaluatedArguments.push_back(arguments[i]);
    }
    return context->call(name, evaluatedArguments);
}

string FunctionCallExpression::toString()
{
    string result = name + "(";

    for (int i = 0, c = arguments.size(); i < c; i++) {
        result += arguments[i]->toString();
        if (i != c - 1) {
            result += ",";
        }
    }

    result += ")";

    return result;
}

// BinaryOperatorExpression

BinaryOperatorExpression::BinaryOperatorExpression(
    string op_,
    shared_ptr<Expression> left_,
    shared_ptr<Expression> right_
): Expression("(operator)"), op(op_), left(left_), right(right_) {
}

Value BinaryOperatorExpression::evaluate(Context* context)
{
    if (op == "+")
        return left->evaluate(context).add(right->evaluate(context));
    if (op == "-")
        return left->evaluate(context).subtract(right->evaluate(context));
    if (op == "*")
        return left->evaluate(context).multiply(right->evaluate(context));
    if (op == "/")
        return left->evaluate(context).divide(right->evaluate(context));
    if (op == "**")
        return left->evaluate(context).pow(right->evaluate(context));
    if (op == "%")
        return left->evaluate(context).mod(right->evaluate(context));
    if (op == "=")
        return left->evaluate(context).equal(right->evaluate(context));
    if (op == "!=")
        return left->evaluate(context).equal(right->evaluate(context)).logicalNot();
    if (op == ">")
        return left->evaluate(context).greater(right->evaluate(context));
    if (op == ">=")
        return left->evaluate(context).greaterOrEqual(right->evaluate(context));
    if (op == "<")
        return left->evaluate(context).less(right->evaluate(context));
    if (op == "<=")
        return left->evaluate(context).lessOrEqual(right->evaluate(context));
    if (op == "&&")
        return left->evaluate(context).logicalAnd(right->evaluate(context));
    if (op == "||")
        return left->evaluate(context).logicalOr(right->evaluate(context));
    if (op == "^^")
        return left->evaluate(context).logicalXor(right->evaluate(context));
    if (op == "&")
        return left->evaluate(context).bitwiseAnd(right->evaluate(context));
    if (op == "|")
        return left->evaluate(context).bitwiseOr(right->evaluate(context));
    if (op == "^")
        return left->evaluate(context).bitwiseXor(right->evaluate(context));
    if (op == ":=") {
        Value assignment = right->evaluate(context);
        if (left->type == "(variable)") {
            shared_ptr<VariableExpression> variable = dynamic_pointer_cast<VariableExpression>(left);
            context->assign(variable->name, assignment);
            return assignment;
        } else if (left->type == "(dynamic_variable)") {
            shared_ptr<DynamicVariableExpression> variable = dynamic_pointer_cast<DynamicVariableExpression>(left);
            context->assign(variable->getName(context), assignment);
            return assignment;
        }
        cerr << "context: invalid left operand for assign: " << left->type << endl;
        return Value();
    }
    cerr << "invalid operator: " << op << endl;
    return 0;
}

string BinaryOperatorExpression::toString()
{
    return left->toString() + op + right->toString();
}

// UnaryOperatorExpression

UnaryOperatorExpression::UnaryOperatorExpression(
    string op_,
    shared_ptr<Expression> right_
): Expression("(operator)"), op(op_), right(right_) {
}

Value UnaryOperatorExpression::evaluate(Context* context)
{
    if (op == "+")
        return right->evaluate(context);
    if (op == "-")
        return right->evaluate(context).inverse();
    if (op == "~")
        return right->evaluate(context).bitwiseNot();
    if (op == "!")
        return right->evaluate(context).logicalNot();
    cerr << "Invalid operator: " << op << endl;
    return Value(0);
}

string UnaryOperatorExpression::toString()
{
    return op + right->toString();
}

// IntervalOperatorExpression

IntervalOperatorExpression::IntervalOperatorExpression(
    string op_,
    string startOp_,
    string endOp_,
    shared_ptr<Expression> value_,
    Value left_,
    Value right_
): Expression("(interval)"), op(op_), startOp(startOp_), endOp(endOp_), value(value_), left(left_), right(right_) {
}

Value IntervalOperatorExpression::evaluate(Context* context)
{
    Value evaluation = value->evaluate(context);
    Value afterStart(0);
    Value beforeEnd(0);

    if (startOp == "[") {
        afterStart = evaluation.greaterOrEqual(left);
    }

    if (startOp == "(") {
        afterStart = evaluation.greater(left);
    }

    if (endOp == "]") {
        beforeEnd = evaluation.lessOrEqual(right);
    }

    if (endOp == ")") {
        beforeEnd = evaluation.less(right);
    }

    Value inRange = afterStart.logicalAnd(beforeEnd);

    return op == "=" ? inRange : inRange.logicalNot();
}

string IntervalOperatorExpression::toString()
{
    string start = left.isString() ? "\"" + left.stringValue() + "\"" : left.stringValue();
    string end = right.isString() ? "\"" + right.stringValue() + "\"" : right.stringValue();
    return startOp + start + "," + end + endOp;
}

// RedirectionExpression

RedirectionExpression::RedirectionExpression(shared_ptr<Expression> left_, shared_ptr<Expression> right_): Expression("(redirection)"), left(left_), right(right_)
{
}

Value RedirectionExpression::evaluate(Context* context)
{
    string contextName;

    if (left->type == "(variable)") {
        contextName = dynamic_pointer_cast<VariableExpression>(left)->name;
    } else if (left->type == "(dynamic_variable)") {
        contextName = dynamic_pointer_cast<DynamicVariableExpression>(left)->getName(context);
    } else {
        return Value();
    }

    shared_ptr<Context> redirect = context->redirect(contextName);

    if (redirect == NULL) {
        return Value();
    }

    return right->evaluate(redirect.get());
}

string RedirectionExpression::toString()
{
    return "(" + left->toString()  + "," + right->toString() + ")";
}

// ArrayExpression

ArrayExpression::ArrayExpression(vector<shared_ptr<Expression>> exps): Expression("(array)"), expressions(exps)
{
}

Value ArrayExpression::evaluate(Context* context)
{
    vector<Value> values;

    for (int i = 0; i < expressions.size(); i++) {
        values.push_back(expressions[i]->evaluate(context));
    }

    return Value(values);
}

string ArrayExpression::toString()
{
    string result = "";

    for (int i = 0, c = expressions.size(); i < c; i++) {
        result += expressions[i]->toString();
        if (i != c - 1) {
            result += ",";
        }
    }

    return result;
}

File: source/native/src/expression/expression.rs
use std::rc::Rc;
use gdnative::prelude::*;
use crate::expression::value::Value;
use crate::expression::context::Context;

pub enum Expression {
    BottomExpression,
    LiteralExpression(Value),
    VariableExpression(String),
    DynamicVariableExpression(String, Rc<Expression>),
    FunctionCallExpression(String, Vec<Rc<Expression>>),
    BinaryOperatorExpression(String, Rc<Expression>, Rc<Expression>),
    UnaryOperatorExpression(String, Rc<Expression>),
    IntervalOperatorExpression(String, String, String, Rc<Expression>, Value, Value),
    RedirectionExpression(Rc<Expression>, Rc<Expression>),
    ArrayExpression(Vec<Rc<Expression>>),
}

impl Expression {
    pub fn to_string(&self) -> String {
        match self {
            Expression::BottomExpression => "(bottom)".to_string(),
            Expression::LiteralExpression(value) => {
                if value.is_string() {
                    return format!("\"{}\"", value.to_string());
                }

                value.to_string()
            },
            Expression::VariableExpression(name) => name.to_string(),
            Expression::DynamicVariableExpression(base_name, argument) => {
                format!("{}({})", base_name, argument.to_string())
            },
            Expression::FunctionCallExpression(name, expressions) => {
                let argument_names: Vec<String> = expressions.iter()
                    .map(| expression | { expression.to_string() })
                    .collect();

                format!("{}({})", name, argument_names.join(","))
            },
            Expression::BinaryOperatorExpression(op, left, right) => {
                format!("{}{}{}", left.to_string(), op, right.to_string())
            },
            Expression::UnaryOperatorExpression(op, right) => {
                format!("{}{}", op, right.to_string())
            },
            Expression::IntervalOperatorExpression(_, start_op, end_op, _, left, right) => {
                let start = if left.is_string() { format!("\"{}\"", left.to_string()) } else { left.to_string() };
                let end = if right.is_string() { format!("\"{}\"", right.to_string()) } else { right.to_string() };
                return format!("{}{},{}{}", start_op, start, end, end_op)
            },
            Expression::RedirectionExpression(left, right) => {
               format!("({},{})", left.to_string(), right.to_string())
            },
            Expression::ArrayExpression(expressions) => {
                let argument_names: Vec<String> = expressions.iter()
                    .map(| expression | { expression.to_string() })
                    .collect();

                argument_names.join(",")
            },
        }
    }

    pub fn get_type_name(&self) -> String {
        match self {
            Expression::BottomExpression => "(bottom)".to_string(),
            Expression::LiteralExpression(_) => "(literal)".to_string(),
            Expression::VariableExpression(_) => "(variable)".to_string(),
            Expression::DynamicVariableExpression(_, _) => "(dynamic_variable)".to_string(),
            Expression::FunctionCallExpression(_, _) => "(call)".to_string(),
            Expression::BinaryOperatorExpression(_, _, _) => "(operator)".to_string(),
            Expression::UnaryOperatorExpression(_, _) => "(operator)".to_string(),
            Expression::IntervalOperatorExpression(_, _, _, _, _, _) => "(interval)".to_string(),
            Expression::RedirectionExpression(_, _) => "(redirection)".to_string(),
            Expression::ArrayExpression(_) => "(array)".to_string(),
        }
    }

    pub fn evaluate(&self, context: &Context) -> Value {
        match self {
            Expression::BottomExpression => Value::BottomValue,
            Expression::LiteralExpression(value) => value.clone(),
            Expression::VariableExpression(name) => context.get(&name),
            Expression::DynamicVariableExpression(base_name, argument) => {
                let name = &[base_name, &argument.evaluate(context).to_string()[..]].join(".");
                context.get(name)
            },
            Expression::FunctionCallExpression(name, expressions) => {
                context.call(name, &expressions)
            },
            Expression::BinaryOperatorExpression(op, left, right) => {
                match &op[..] {
                    "+" => left.evaluate(context).add(&right.evaluate(context)),
                    "-" => left.evaluate(context).subtract(&right.evaluate(context)),
                    "*" => left.evaluate(context).multiply(&right.evaluate(context)),
                    "/" => left.evaluate(context).divide(&right.evaluate(context)),
                    "**" => left.evaluate(context).pow(&right.evaluate(context)),
                    "%" => left.evaluate(context).modl(&right.evaluate(context)),
                    "=" => left.evaluate(context).equal(&right.evaluate(context)),
                    "!=" => left.evaluate(context).equal(&right.evaluate(context)).logical_not(),
                    ">" => left.evaluate(context).greater(&right.evaluate(context)),
                    ">=" => left.evaluate(context).greater_or_equal(&right.evaluate(context)),
                    "<" => left.evaluate(context).less(&right.evaluate(context)),
                    "<=" => left.evaluate(context).less_or_equal(&right.evaluate(context)),
                    "&&" => left.evaluate(context).logical_and(&right.evaluate(context)),
                    "||" => left.evaluate(context).logical_or(&right.evaluate(context)),
                    "^^" => left.evaluate(context).logical_xor(&right.evaluate(context)),
                    "&" => left.evaluate(context).bitwise_and(&right.evaluate(context)),
                    "|" => left.evaluate(context).bitwise_or(&right.evaluate(context)),
                    "^" => left.evaluate(context).bitwise_xor(&right.evaluate(context)),
                    ":=" => {
                        let assignment = right.evaluate(context);
                        let left_deref: &Expression = &**left;
                        if let Expression::VariableExpression(name) = left_deref {
                            context.assign(&name, &assignment);
                            return assignment;
                        }
                        if let Expression::DynamicVariableExpression(base_name, argument) = left_deref {
                            let name = &[base_name, &argument.evaluate(context).to_string()[..]].join(".");
                            context.assign(name, &assignment);
                            return assignment;
                        }
                        godot_print!("context: invalid left operand for assign");
                        Value::BottomValue
                    },
                    _ => Value::BottomValue,
                }
            },
            Expression::UnaryOperatorExpression(op, right) => {
                match &op[..] {
                    "+" => right.evaluate(context),
                    "-" => right.evaluate(context).inverse(),
                    "~" => right.evaluate(context).bitwise_not(),
                    "!" => right.evaluate(context).logical_not(),
                    _ => Value::BottomValue,
                }
            },
            Expression::IntervalOperatorExpression(op, start_op, end_op, expression, left, right) => {
                let evaluation = expression.evaluate(context);
                let mut after_start = Value::IntValue(0);
                let mut before_end = Value::IntValue(0);

                if start_op == "[" {
                    after_start = evaluation.greater_or_equal(left);
                }

                if start_op == "(" {
                    after_start = evaluation.greater(left);
                }

                if end_op == "]" {
                    before_end = evaluation.less_or_equal(right);
                }

                if end_op == ")" {
                    before_end = evaluation.less(right);
                }

                let in_range = after_start.logical_and(&before_end);

                if op == "=" { in_range } else { in_range.logical_not() }
            },
            Expression::RedirectionExpression(left, right) => {
                let context_name: String;
                let left_deref = &**left;

                if let Expression::VariableExpression(name) = left_deref {
                    context_name = name.to_string();
                } else if let Expression::DynamicVariableExpression(base_name, argument) = left_deref {
                    context_name = [base_name, &argument.evaluate(context).to_string()[..]].join(".");
                } else {
                    return Value::BottomValue;
                }

                match context.redirect(&context_name) {
                    Option::Some(redirected_context) => right.evaluate(&redirected_context),
                    Option::None => Value::BottomValue
                }
            },
            Expression::ArrayExpression(expressions) => {
                let mut values: Vec<Value> = Vec::new();
                for expression in expressions.iter() {
                    values.push(expression.evaluate(context))
                }
                return Value::ArrayValue(values)
            },
        }
    }
}
